pallets/auctions/src/types.rs

```diff
@@ -0,0 +1,246 @@
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+
+use codec::{Decode, Encode};
+use scale_info::TypeInfo;
+use sp_runtime::RuntimeDebug;
+use ternoa_primitives::marketplace::MarketplaceId;
+
+#[derive(Encode, Decode, Clone, PartialEq, RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+/// Structure to store Auction data
+pub struct AuctionData<AccountId, BlockNumber, BalanceCaps>
+where
+    AccountId: Clone + Default,
+    BalanceCaps: Clone + Default,
+{
+    pub creator: AccountId,
+    pub start_block: BlockNumber,
+    pub end_block: BlockNumber,
+    pub start_price: BalanceCaps,
+    pub buy_it_price: Option<BalanceCaps>,
+    pub bidders: SortedBidderList<AccountId, BalanceCaps>,
+    pub marketplace_id: MarketplaceId,
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+/// wrapper type to store sorted list of all bids
+/// The wrapper exists to ensure a queue implementation of sorted bids
+pub struct SortedBidderList<AccountId, BalanceCaps>(pub Vec<(AccountId, BalanceCaps)>);
+
+impl<AccountId, BalanceCaps> SortedBidderList<AccountId, BalanceCaps>
+where
+    AccountId: std::cmp::Ord + Clone,
+    BalanceCaps: std::cmp::PartialOrd,
+{
+    pub const MAX_COUNT: usize = 10;
+
+    /// Create a new empty bidders list
+    pub fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    /// Insert a new bid to the list
+    ///
+    /// The function ensures that the new bid is always higher than the current highest bid
+    /// If the insert action will cause an overflow, the lowest bid is removed and returned
+    /// If value lower than current highest bid is passed, the function will panic
+    pub fn insert_new_bid(
+        &mut self,
+        account_id: AccountId,
+        value: BalanceCaps,
+    ) -> Option<(AccountId, BalanceCaps)> {
+        // ensure the new bid is larger than current highest bid
+        if let Some(current_highest_bid) = self.get_highest_bid() {
```

Comment:
To continue the discussion....
In this (rare) case, the caller should be responsible for checking if the input values are correct.
With your implementation we are doing the same check twice and you have open this function to panic which should never be the case.

Reply:
The general rule is to separate the checking part from doing the actual work (if possible) and the check has been already done.

Reply:
Also if you remove this check, you will have an interface that does what it is supposed to do without causing any side effects.

Reply:
done, removed the check and  I've renamed it to BiddersList